src/middleware/setup.ts

```diff
@@ -0,0 +1,41 @@
+import {
+  GetServerSidePropsContext,
+  NextApiHandler,
+  NextApiRequest,
+  NextApiResponse,
+} from "next";
+import { SetupMiddlewareArgs } from "../types";
+import { tokens } from "../csrf";
+import { sign } from "cookie-signature";
+import { serialize } from "cookie";
+
+type SetupArgs =
+  | NextApiRequest[]
+  | NextApiResponse[]
+  | GetServerSidePropsContext[];
+
+const setup = (
+  handler: NextApiHandler,
+  { csrfSecret, secret, tokenKey, cookieOptions }: SetupMiddlewareArgs
+) => async (...args: SetupArgs): Promise<void> => {
+  const isApi = args.length > 1;
+
+  const req = isApi
+    ? (args[0] as NextApiRequest) // (*req*, res)
+    : (args[0] as GetServerSidePropsContext).req; // (context).req
+  const res = isApi
+    ? (args[1] as NextApiResponse) // (req, *res*)
+    : (args[0] as GetServerSidePropsContext).res; // (context).res
```

Comment:
I think setup function should receives `{ req, res }` object.

Reply:
As part of the `SetupMiddlewareArgs`? Can you elaborate a bit more on your suggestion? Thanks by the way.